Sources/App/Commands/TriggerBuilds.swift

```diff
@@ -184,68 +173,76 @@ func triggerBuilds(on database: Database,
 ///   - logger: `Logger` used for logging
 ///   - packages: list of `Package.Id`s to trigger
 ///   - force: do not check pipeline capacity and ignore downscaling
-/// - Returns: `EventLoopFuture<Void>` future
 func triggerBuilds(on database: Database,
                    client: Client,
                    logger: Logger,
                    packages: [Package.Id],
-                   force: Bool = false) -> EventLoopFuture<Void> {
+                   force: Bool = false) async throws {
     guard Current.allowBuildTriggers() else {
         logger.info("Build trigger override switch OFF - no builds are being triggered")
-        return database.eventLoop.future()
+        return
     }
 
     guard !force else {
-        return packages.map {
-            findMissingBuilds(database, packageId: $0).flatMap {
-                triggerBuildsUnchecked(on: database, client: client, logger: logger, triggers: $0) }
+        return await withThrowingTaskGroup(of: Void.self) { group in
+            for package in packages {
+                group.addTask {
+                    let triggerInfo = try await findMissingBuilds(database, packageId: package)
+                    try await triggerBuildsUnchecked(on: database, client: client, logger: logger, triggers: triggerInfo)
+                }
+            }
         }
-        .flatten(on: database.eventLoop)
     }
 
-    return Current.getStatusCount(client, .pending)
-        .and(Current.getStatusCount(client, .running))
-        .flatMap { (pendingJobs, runningJobs) in
-            AppMetrics.buildPendingJobsCount?.set(pendingJobs)
-            AppMetrics.buildRunningJobsCount?.set(runningJobs)
-            var newJobs = 0
-            return packages.map { pkgId in
-                let allowListed = Current.buildTriggerAllowList().contains(pkgId)
-                let downscalingAccepted = Current.random(0...1) < Current.buildTriggerDownscaling()
-                guard allowListed || downscalingAccepted else {
-                    logger.info("Build trigger downscaling in effect - skipping builds")
-                    return database.eventLoop.future()
-                }
+    async let pendingJobsTask = Current.getStatusCount(client, .pending).get()
+    async let runningJobsTask = Current.getStatusCount(client, .running).get()
+    let pendingJobs = try await pendingJobsTask
+    let runningJobs = try await runningJobsTask
+
+    AppMetrics.buildPendingJobsCount?.set(pendingJobs)
+    AppMetrics.buildRunningJobsCount?.set(runningJobs)
+
+    let newJobs = ActorIsolated(0)
+
+    await withThrowingTaskGroup(of: Void.self) { group in
```

Comment:
Same here: double check concurrency behaviour. We don't want to be flooding Gitlab.